prototypes/matrix_free_stokes/matrix_free_stokes.cc

```diff
@@ -0,0 +1,2102 @@
+
+  std::vector<IndexSet> refinement_edge_indices;
+  refinement_edge_indices.clear();
+  const unsigned int nlevels =
+    matrix_free.get_dof_handler().get_triangulation().n_global_levels();
+  refinement_edge_indices.resize(nlevels);
+  for (unsigned int l = 0; l < nlevels; l++)
+    refinement_edge_indices[l] =
+      IndexSet(matrix_free.get_dof_handler().n_dofs(l));
+
+  MGTools::extract_inner_interface_dofs(matrix_free.get_dof_handler(),
+                                        refinement_edge_indices);
+  return refinement_edge_indices[level];
+}
+
+// Multigrid global coarsening algorithm and parameters
+struct MultigridParameters
+{
+  struct
+  {
+    std::string  type            = "gmres_with_amg";
+    unsigned int maxiter         = 2000;
+    double       abstol          = 1e-14;
+    double       reltol          = 1e-4;
+    unsigned int smoother_sweeps = 1;
+    unsigned int n_cycles        = 1;
+    std::string  smoother_type   = "ILU";
+  } coarse_solver;
+
+  struct
+  {
+    std::string  type         = "relaxation";
+    unsigned int n_iterations = 10;
+    double       relaxation   = 0.5;
+  } smoother;
+};
+
+template <typename VectorType,
+          int dim,
+          typename SystemMatrixType,
+          typename LevelMatrixType,
+          typename MGTransferType>
+static void
+gcmg_solve(SolverControl &            solver_control,
+           VectorType &               dst,
+           const VectorType &         src,
+           const MultigridParameters &mg_data,
+           const DoFHandler<dim> &    dof,
+           const SystemMatrixType &   fine_matrix,
+           const MGLevelObject<std::unique_ptr<LevelMatrixType>> &mg_matrices,
+           const MGTransferType &                                 mg_transfer)
+{
+  AssertThrow(mg_data.coarse_solver.type == "gmres_with_amg",
+              ExcNotImplemented());
+  AssertThrow(mg_data.smoother.type == "relaxation", ExcNotImplemented());
+
+  const unsigned int min_level = mg_matrices.min_level();
+  const unsigned int max_level = mg_matrices.max_level();
+
+  using SmootherPreconditionerType = DiagonalMatrix<VectorType>;
+  using SmootherType =
+    PreconditionRelaxation<LevelMatrixType, SmootherPreconditionerType>;
+  using PreconditionerType = PreconditionMG<dim, VectorType, MGTransferType>;
+
+  mg::Matrix<VectorType> mg_matrix(mg_matrices);
+
+  MGLevelObject<typename SmootherType::AdditionalData> smoother_data(min_level,
+                                                                     max_level);
+
+  for (unsigned int level = min_level; level <= max_level; ++level)
+    {
+      smoother_data[level].preconditioner =
+        std::make_shared<SmootherPreconditionerType>();
+      mg_matrices[level]->compute_inverse_diagonal(
+        smoother_data[level].preconditioner->get_vector());
+      smoother_data[level].n_iterations = mg_data.smoother.n_iterations;
+      smoother_data[level].relaxation   = mg_data.smoother.relaxation;
```

Comment:
How do you select the value? You could automated it via https://github.com/peterrum/dealii-dd-and-schwarz/blob/d998b9b344a19c9d2890e087f953c2f93e6546ae/include/precondition.templates.h#L294-L309.

Reply:
Nice, so far we have been playing around with it without automating it. I think we will leave the prototype like this and then experiment with what you suggested later on...